crates/amaru-ledger/src/rules/transaction/vkey_witness.rs

```diff
@@ -0,0 +1,204 @@
+                        }
+                    }
+                    Certificate::UnReg(stake_credential, _) => stake_credential.get_key_hash(),
+                    Certificate::VoteDeleg(stake_credential, _) => stake_credential.get_key_hash(),
+                    Certificate::StakeVoteDeleg(stake_credential, _, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::StakeRegDeleg(stake_credential, _, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::VoteRegDeleg(stake_credential, _, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::StakeVoteRegDeleg(stake_credential, _, _, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::AuthCommitteeHot(stake_credential, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::ResignCommitteeCold(stake_credential, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::RegDRepCert(stake_credential, _, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::UnRegDRepCert(stake_credential, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                    Certificate::UpdateDRepCert(stake_credential, _) => {
+                        stake_credential.get_key_hash()
+                    }
+                })
+                .collect::<Vec<_>>()
+        })
+        .unwrap_or_default();
+
+    let required_vkey_hashes = [
+        spend_pkhs.as_slice(),
+        required_signers.as_slice(),
+        withdrawal_pkhs.as_slice(),
+        vote_pkhs.as_slice(),
+        certificate_pkhs.as_slice(),
+    ]
+    .concat();
+
+    let empty_vec = vec![];
+    let vkey_witnesses = vkey_witnesses.as_deref().unwrap_or(&empty_vec);
+    let vkey_hashes = vkey_witnesses
+        .iter()
+        .map(|witness| Hasher::<224>::hash(&witness.vkey))
+        .collect::<Vec<_>>();
+
+    // Are we worried about efficiency here? this is quadratic time
+    let missing_key_hashes: Vec<_> = required_vkey_hashes
+        .into_iter()
+        .filter(|hash| !vkey_hashes.contains(hash))
+        .collect();
+
+    if !missing_key_hashes.is_empty() {
+        return Err(TransactionRuleViolation::MissingRequiredVkeyWitnesses { missing_key_hashes });
+    }
+
+    let invalid_witnesses = vkey_witnesses
+        .iter()
+        .filter(|witness| {
+            match validate_witness(witness, transaction_body.original_hash().as_slice()) {
+                Ok(is_valid) => !is_valid,
+                Err(e) => {
+                    eprintln!("Failed to validate witness: {:?}", e);
```

Comment:
Same as in `bootstrap_witness`,  I don't think this log line is helpful

Reply:
I just wanted to avoid swallowing the error and providing no context as to why it failed.

Reply:
You might want to introduce `tracing` then.